### DISCUSSION TITLE
Homer 10 CDR

### DISCUSSION THREAD

**Discussion by SerhiiZahuba** [sipcapture/homer#620] [General]
Who has the opportunity to give advice? I want to see CDR in Homer 10 just like in this screenshot. Unfortunately, I did not find documentation on setting this up. I use asterisk pjsip + configured in asterisk hep.conf. As far as I understand, everything is configured in HOmer and you need to send data from asterisk there correctly.
1) It is necessary to install the heplify agent, will what is already transmitted by asterisk via hep.conf be suitable?
2) What exactly needs to be added to the dialplan parameters to display data?
3) I also see that the cost per call can be displayed on the graph - where can this be done?

Until I found the answer to these questions. If someone can help, I will be very grateful.

![image](https://github.com/sipcapture/homer/assets/36307887/39f2d8e7-e320-4664-92d4-369245633af7)

**Comment by lmangani** [comment]
@SerhiiZahuba what is your setup exactly? this example is part of the HOMER 10 demo, on purpose, to demonstrate how it works.

**Comment by SerhiiZahuba** [comment]
I understand. For demo I see they use hepgen to generate calls. I want to understand how to do the same in myself and get such visualization with asterisk.

**Comment by lmangani** [comment]
The best way to start is by copying the example and sending your CDRs in as JSON objects into qryn so you can then learn how to parse them with LogQL using the provided examples which show how to extract from those objects. Of course the quality of the input determines the quality of the output.

**Reply by pahaz** [reply]
Do you have any examples for Asterisk or FreeSwitch?

**Comment by tony1661** [comment]
Hi, I am also trying to understand how CDRs work in Homer 10.

@lmangani you mentioned above that we need to send our logs into qryn as a JSON object.

Just so I understand, would I need to extract the CDRs from my system via a script or something like that? Or, will heplify or captagent do that for me?

I was (maybe incorrectly) under the assumption that a capture agent would read the SIP messages and be able to generate the CDRs.

Thanks 😃

**Comment by adubovikov** [comment]
Hi all,

This thread mixes two different things. Worth separating them.

### What the Homer 10 screenshot actually is

That dashboard is from the **Homer 10 demo**, fed by synthetic CDR JSON (`hepgen`), not by Asterisk `hep.conf`.

The demo flow was:

1. Generate or export CDR records as **JSON objects**
2. Push them into **qryn**
3. Build Grafana panels with **LogQL** to extract fields (duration, cost, MOS, etc.)

So the screenshot is a **visualization example**, not something Asterisk HEP mirroring turns on by itself.

### What Asterisk `hep.conf` gives you

`res_hep` / `res_hep_pjsip` send **SIP signaling** over HEPv3.

That is useful for call tracing (INVITE/BYE, responses, Call-ID, flow), but it is **not** Asterisk billing CDR.

No dialplan parameter will make `hep.conf` emit:

- billable duration
- per-call cost/rate
- carrier billing fields
- custom accounting columns from `cdr.conf`

Those live in Asterisk's own CDR engine (`cdr_csv`, `cdr_manager`, ODBC/PostgreSQL CDR, etc.).

### Do heplify / captagent generate CDR?

**Not billing CDR.**

Capture agents mirror/decode SIP (and related protocols) from the network or from HEP input. From SIP alone you can derive **signaling-side timing** (setup, ringing, answered, release), but not tariff/cost unless you add that data yourself.

So [@tony1661](https://github.com/tony1661): your assumption is partly right for **call timing from SIP**, but wrong for **billing CDR dashboards** like the demo.

### How to get CDR-like data in practice

**Option A — SIP-derived call info (simplest with Asterisk + hep.conf)**

Keep sending SIP via HEP. Homer derives call duration / ringing / status from SIP messages (INVITE, 18x, 200, BYE).

Good for troubleshooting and basic KPIs. No cost field unless you add it elsewhere.

**Option B — Export real CDR from Asterisk/FreeSWITCH (for billing fields, cost, custom columns)**

Use the switch CDR export path, then forward JSON records to your observability stack:

- **Asterisk:** `cdr_manager`, `cdr_adaptive_odbc`, post-call AGI/AMI script, or file watcher on CSV CDR
- **FreeSWITCH:** mod_json_cdr / XML CDR export

On **Homer 10 preview**: push those JSON objects into qryn and parse with LogQL (as [@lmangani](https://github.com/lmangani) noted).

On **Homer 11 (current)**: primary path is SIP trace + correlation in the bundled UI. Full external billing CDR ingest (telecom-style records, cost, MOS from SBC/RADIUS) is a separate integration layer; we use dedicated collectors for RADIUS/FTP CDR from network gear, not from `hep.conf`.

There is **no official Asterisk/FreeSWITCH CDR → Homer recipe** in this repo today. Typical approach is a small script/service that reads Asterisk CDR events and publishes JSON to your ingest endpoint.

**Option C — Network/SBC CDR**

If you have an SBC/gateway emitting RADIUS Accounting or CDR files, a CDR collector agent can normalize and forward those records upstream. That is the path for carrier-grade fields (ingress/egress, MOS, billing duration, etc.).

### About "cost per call" on the graph

That was **demo data** in the sample JSON. Homer does not calculate call cost from SIP. You must supply `cost` (or equivalent) in the CDR JSON from your billing/rating system.

### Where to start today

| Goal | Setup |
|------|--------|
| SIP tracing from Asterisk | `hep.conf` + `res_hep_pjsip` → heplify/hepagent → Homer |
| Demo-style CDR dashboard (H10 preview) | export JSON CDR → qryn → Grafana/LogQL |
| Billing CDR with cost | Asterisk/FS CDR export + custom forwarder; quality of dashboard = quality of your JSON |
| Production today | **Homer 11** for SIP capture/search; treat H10 CDR dashboard as preview pattern |

If someone wants to wire Asterisk `cdr_manager` → JSON → Homer 11 ingest, open a new thread with your Asterisk version and preferred CDR backend (csv/manager/odbc). Happy to sketch the pipeline.

[@pahaz](https://github.com/pahaz): no maintained Asterisk/FreeSWITCH example in-tree yet; the pattern above is what teams use in the field.
